### PR TITLE
Adds image cache images to build

### LINKED ISSUE #2918: Builds should have access to previous builds' images
Builds should have access to previous builds' images if they are available.
This functionality can be particularly useful if one is interested in reusing previously built assets etc. from previous deployments.

### PR DESCRIPTION
**Pull request by bomoko**
This PR introduces a new set of build args made available to the present build that give precise reference to the images rolled out in the last deployment.

It interrogates the current deployment list and introduces build args `LAGOON_CACHE_{service name}` for each service.
The service name is drawn by matching any images in the deployment with the standard structure `harborbase/{project}/{environment}/{service}/...`.

# Checklist

- [ ] Affected Issues have been mentioned in the Closing issues section
- [ ] Documentation has been written/updated
- [ ] PR title is ready for changelog and subsystem label(s) applied

Explain the **details** for making this change. What existing problem does the pull request solve?

Details for what issue this solves are described in the linked issue

# Closing issues

closes #2918

### CODE REVIEW

**Comment by tobybellwood** [comment]
maybe we can get a brief para in the docs about this too
https://docs.lagoon.sh/lagoon/using-lagoon-the-basics/build-and-deploy-process#3.-build-image

**Comment by bomoko** [comment]
100% @tobybellwood - I think that once we've done some testing and tweaking on the test cluster, I'll do a writeup.

**Comment by bryangruneberg** [comment]
yes! this is _nice_

File: images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh [hunk 1]
```diff
@@ -413,6 +413,15 @@ do
 
 done
 
+##############################################
+### CACHE IMAGE LIST GENERATION
+##############################################
+
+LAGOON_CACHE_BUILD_ARGS=()
+readarray LAGOON_CACHE_BUILD_ARGS < <(kubectl --insecure-skip-tls-verify -n ${NAMESPACE} get deployments -y yaml | yq e '.items[].spec.template.spec.containers[].image | capture("^(?P<image>.+\/.+\/.+\/(?P<name>.+)\@.*)$") | "LAGOON_CACHE_" + .name + "=\"" + .image + "\""' -)
```
Comment (bomoko) [inline images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh]: Filter deployments

File: images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh [hunk 2]
```diff
@@ -462,6 +462,15 @@ patchBuildStep "${buildStartTime}" "${buildStartTime}" "${currentStepEnd}" "${NA
 previousStepEnd=${currentStepEnd}
 set -x
 
+##############################################
+### CACHE IMAGE LIST GENERATION
+##############################################
+
+LAGOON_CACHE_BUILD_ARGS=()
+readarray LAGOON_CACHE_BUILD_ARGS < <(kubectl --insecure-skip-tls-verify -n ${NAMESPACE} get deployments -o yaml -l 'lagoon.sh/service' | yq e '.items[].spec.template.spec.containers[].image | capture("^(?P<image>.+\/.+\/.+\/(?P<name>.+)\@.*)$") | "LAGOON_CACHE_" + .name + "=\"" + .image + "\""' -)
```
Comment (tobybellwood) [inline images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh:470]: we're getting this erroring on our version of yq
```
ci-nginx-control-k8s-nginx lagoon-build-72o75 jenkinslagoonpr29197-control-plane lagoon-build 2021-11-23T23:43:25.995103783Z Error: unknown command "e" for "yq"
```

exit status only appeared in yq 4 -we'll need to upgrade the version we use here (our version = 3.3.2, current release [4.15.1](https://github.com/mikefarah/yq/releases)

Comment (bomoko) [reply images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh:470]: @tobybellwood should we upgrade in this PR (can we?)

Comment (bomoko) [reply images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh:470]: I can probably get it to work without it as well (perhaps that's better?)

Comment (tobybellwood) [reply images/kubectl-build-deploy-dind/build-deploy-docker-compose.sh:470]: maybe for now, working without yq4, as it's a bit of an upgrade.

Or do a second `yq4` binary, and we can switch the rest of the yq logic across later?

**Comment by tobybellwood** [comment]
We need a test, and some docs for this, ping @bomoko to follow up

**Pull request merged**